### PR TITLE
[WIP] Fast Preview

### PR DESCRIPTION
**Pull request by jasonLaster**
Associated Issue: #5377 

### Summary of Changes

* Early hacky version. Give it a go and let me what you think.
* It somewhat works... 
  - finds lots of expression
  - highlighting is off
  - is flickery

### CODE REVIEW

File: src/actions/preview.js [hunk 1]
```diff
@@ -61,16 +62,43 @@ async function getExtraProps(expression, result, evaluate) {
...
+function isInvalidTarget(target: HTMLElement) {
+  if (!target || !target.innerText) {
+    return true;
+  }
+
+  const tokenText = target.innerText.trim();
+  const cursorPos = target.getBoundingClientRect();
+
+  // exclude literal tokens where it does not make sense to show a preview
+  const invaildType = ["cm-string", "cm-number", "cm-atom"].includes(
```
Comment (codehag) [inline src/actions/preview.js]: "invalid"

File: src/actions/preview.js [hunk 2]
```diff
@@ -61,16 +62,43 @@ async function getExtraProps(expression, result, evaluate) {
...
+    const tokenPos = getTokenLocation(editor.codeMirror, target);
     const cursorPos = target.getBoundingClientRect();
     const preview = getPreview(getState());
 
     if (preview) {
-      // We are mousing over the same token as before
-      if (isEqual(preview.tokenPos, location)) {
+      // Return early if we are currently showing another preview or
+      // if we are mousing over the same token as before
+      if (preview.updating || isEqual(preview.tokenPos, tokenPos)) {
```
Comment (codehag) [inline src/actions/preview.js]: `preview.updating` might be why we are seeing the flicker

File: src/actions/preview.js [hunk 3]
```diff
@@ -80,72 +108,44 @@ export function updatePreview(target: HTMLElement, editor: any) {
       }
     }
 
-    const source = getSelectedSource(getState());
-
-    const symbols = getSymbols(getState(), source.toJS());
-    if (symbols.identifiers.length == 0) {
+    if (isInvalidTarget(target)) {
```
Comment (codehag) [inline src/actions/preview.js]: :+1: nice change

File: src/actions/preview.js [hunk 4]
```diff
@@ -80,72 +108,44 @@ export function updatePreview(target: HTMLElement, editor: any) {
...
-
-    const isUpdating = preview && preview.updating;
+    if (!isLineInScope(getState(), tokenPos.line)) {
+      return;
+    }
 
-    const linesInScope = getInScopeLines(getState());
-    const inScope = linesInScope && linesInScope.includes(location.line);
+    const source = getSelectedSource(getState());
+    const symbols = getSymbols(getState(), source.toJS());
```
Comment (codehag) [inline src/actions/preview.js]: hmmm im a bit surprised that getSymbols needs `toJS` -- is it because its used outside of the parser in this case?

**Comment by jasonLaster** [comment]
the flicker was fixed. that was coming from a timing issue :)

File: src/actions/preview.js (hunk 4 above)
Comment (jasonLaster) [inline src/actions/preview.js]: the toJS is on the source

File: src/actions/preview.js (hunk 2 above)
Comment (jasonLaster) [inline src/actions/preview.js]: the flicker was a timing issue that is fixed in this version

**Comment by GCorbel** [comment]
I did some tests with your branch and it seems to very light and sources seems to appears instantly. As you can see [in this memory report](https://github.com/devtools-html/debugger.html/files/1725613/memory-report.json.gz), a single tab opened with the debugger took only 271mb in the "explicit allocations" compared to 320 in the nightly release.

As I said [in my bug report](https://bugzilla.mozilla.org/show_bug.cgi?id=1432051#c14), speed was really improved in latest releases so it's a bit hard to know if this PR does a real difference or not. I attach a report to the comment.

Is it something I should check to see if there is a real difference? Is there a way to be sure I'm using Firefox with your version of the code ?

Thanks a lot for your work.

**Comment by GCorbel** [comment]
I just did a few more tests and I also have 270mb with Firefox nightly so I really don't know there is some improvement.

**Comment by jasonLaster** [comment]
that's fine. i dont think this will actually help memory, just performance. as in preview should show up instantly

**Comment by GCorbel** [comment]
By preview, you mean the little popup which is supposed to appear when we stop the execution on a line with a break point and we pass hover a variable, right ?

In my case, I simply have nothing. When I pass over a variable, nothing appears.

**Comment by jasonLaster** [comment]
yep. oh weird. mind sharing a screenshot of the code?

**Comment by GCorbel** [comment]
Here is the screenshot.

capture du 2018-02-14 17-55-05

My mouse is hover `mod` and there is something in it as you can see in the console. It does the same on the nightly release and the "normal" one too.

**Comment by jasonLaster** [comment]
strange. thanks

**Comment by GCorbel** [comment]
It works but it takes a while for the debugger to be ready. 

Here is a screenshot with my CPU when open the debugger with **Firefox nightly**, wait for the breakpoint and pass hover a variable : 

image

The graph represents 60 seconds and the  60 is the oldest. At second 40, the debugger is ready. From seconds 40 to 20, I pass hover variables but nothing is shown. At seconds 15, I can see the preview. After that, it continues to have a high memory usage but it works.

Here is the same **with your branch** :

image.

It takes the same time to be ready but the previews are correct very soon an it takes a lot less time to complete the same thing.

My conclusion is : your branch is much more efficient.

**Comment by jasonLaster** [comment]
thanks

**Pull request merged**